Web Inspector: Show warnings in the Font details sidebar for synthesized boldness/obliqueness

patrickangle opened the pull request:
#### 805dec11f1f29a24a3754cf7519c9c06e342ba92
<pre>
Web Inspector: Show warnings in the Font details sidebar for synthesized boldness/obliqueness
<a href="https://bugs.webkit.org/show_bug.cgi?id=244718">https://bugs.webkit.org/show_bug.cgi?id=244718</a>
rdar://99496142

Reviewed by Devin Rousso.

Added test cases to: LayoutTests/inspector/css/getComputedPrimaryFontForNode.html

* LayoutTests/inspector/css/getComputedPrimaryFontForNode.html:
* LayoutTests/inspector/css/getComputedPrimaryFontForNode-expected.txt:

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::buildObjectForFont):
- Send the already existing information about font synthesis to the frontend.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Font.js:
(WI.Font.fromPayload):
(WI.Font.prototype.get synthesizedBold):
(WI.Font.prototype.get synthesizedOblique):

* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section &gt; .content &gt; .group &gt; .row.simple:has(.warning)):
(.details-section &gt; .content &gt; .group &gt; .row.simple &gt; .warning):
* Source/WebInspectorUI/UserInterface/Views/DetailsSectionSimpleRow.js:
(WI.DetailsSectionSimpleRow):
(WI.DetailsSectionSimpleRow.prototype.get warningMessage):
(WI.DetailsSectionSimpleRow.prototype.set warningMessage):
- Add support for putting the row in a warning state.

* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.initialLayout):
- Adopt DetailsSectionWarnableSimpleRow and show warnings when the font is synthesized in some way.

Canonical link: <a href="https://commits.webkit.org/254188@main">https://commits.webkit.org/254188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b60ea5e7d8201a5f52aefeee85e4abb8510df8e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97229 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152729 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92062 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30625 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26555 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80185 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91966 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93703 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24680 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74734 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24648 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67633 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79845 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28254 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13618 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73598 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28348 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14597 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26139 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2945 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37490 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76442 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33831 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16962 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -253,7 +253,9 @@
-                { "name": "variationAxes", "type": "array", "items": { "$ref": "FontVariationAxis" }, "description": "The variation axes defined by the font." }
+                { "name": "variationAxes", "type": "array", "items": { "$ref": "FontVariationAxis" }, "description": "The variation axes defined by the font." },
+                { "name": "syntheticBold", "type": "boolean", "optional": true, "description": "Whether the font has synthesized its boldness or not." },
+                { "name": "syntheticOblique", "type": "boolean", "optional": true, "description": "Whether the font has synthesized its obliqueness or not" }
```
NIT: Maybe `synthesizedBold` and `synthesizedOblique` to make it sound more like a verb instead of a noun (i.e. "what is the synthetic bold" vs "is it a synthesized bold")?

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Font.js:
```diff
@@ -25,24 +25,33 @@
         this._name = name;
         this._variationAxes = variationAxes;
+        this._syntheticBold = syntheticBold;
+        this._syntheticOblique = syntheticOblique;
```
Since these are optional (since they're inside a `{...} = {}`) we should have the fallback behavior be here instead of inside `fromPayload`.
```
this._syntheticBold = !!syntheticBold;
this._syntheticOblique = !!syntheticOblique;
```

also ditto ([Source/JavaScriptCore/inspector/protocol/CSS.json](#diff-e4ce9b37fd298901a223600e617fea37991c6ceaa01e238a5cdb9dd67d7190f8):257)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSectionWarnableSimpleRow.js:
```diff
@@ -0,0 +1,60 @@
+        if (!this._warningElement.parentNode && this._warningMessage)
+            this.element.appendChild(this._warningElement);
+        else
+            this._warningElement.remove()
```
Style: missing `;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSectionWarnableSimpleRow.js:
```diff
@@ -0,0 +1,60 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+WI.DetailsSectionWarnableSimpleRow = class DetailsSectionWarnableSimpleRow extends WI.DetailsSectionSimpleRow
```
Honestly this feels like something we could just have be part of `WI.DetailsSectionSimpleRow`.  I could see this being used in other places in the right circumstances.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSectionWarnableSimpleRow.js:
```diff
@@ -0,0 +1,60 @@
+        this._warningMessage = null;
+
+        this._warningElement = document.createElement("div");
+        this._warningElement.className = "warning";
```
Can we lazily create this?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSectionWarnableSimpleRow.css:
```diff
@@ -0,0 +1,51 @@
+.details-section > .content > .group > .row.simple.warnable > .warning {
+    --warning-icon-size: 11px;
+    --warning-icon-inline-padding: 2px;
+    --warning-badge-arrow-depth: 6px;
```
Style: we normally put CSS variables at the end of rulesets

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DetailsSectionWarnableSimpleRow.css:
```diff
@@ -0,0 +1,51 @@
+    --warning-icon-size: 11px;
+    --warning-icon-inline-padding: 2px;
+    --warning-badge-arrow-depth: 6px;
+
```
Style: We don't normally add newlines in CSS rulesets.

dcrousso commented on LayoutTests/inspector/css/getComputedPrimaryFontForNode-expected.txt:
```diff
@@ -61,3 +66,28 @@ PASS: Width axis minimum should be '0'.
+-- Running test setup.
+PASS: Font should be loaded.
+-- Running test case: CSS.getComputedPrimaryFontForNode.SyntheticWeight1
+PASS: Font should not have a synthesized weight.
```
NIT: It's a bit odd that a test with the name `SyntheticWeight1` doesn't have a synthetic weight.  Perhaps (re)name this something like `CSS.getComputedPrimaryFontForNode.HasWeight` instead?

ditto below

dcrousso commented on LayoutTests/inspector/css/getComputedPrimaryFontForNode.html:
```diff
@@ -163,6 +163,66 @@
+    });
+
+    addTestCase({
+        name: "CSS.getComputedPrimaryFontForNode.SyntheticWeight1",
```
Did you mean to name this `SyntheticOblique`?

ditto below

dcrousso commented on LayoutTests/inspector/css/getComputedPrimaryFontForNode.html:
```diff
@@ -220,6 +287,13 @@
     <div id="variable-font-test-1" style="font-family: VariableFont;">Variable Font Test 1</div>
+
+    <div id="synthetic-weight-test-1" style="font-family: AhemNormal; font-weight: normal;">Synthetic Weight Test 1</div>
+    <div id="synthetic-weight-test-2" style="font-family: AhemNormal; font-weight: bold;">Synthetic Weight Test 2</div>
```
I think more descriptive names here would be great.  Maybe something like `synthetic-weight-from-normal` and `synthetic-weight-from-bold`?

ditto below

patrickangle committed:
Web Inspector: Show warnings in the Font details sidebar for synthesized boldness/obliqueness
https://bugs.webkit.org/show_bug.cgi?id=244718
rdar://99496142

Reviewed by Devin Rousso.

Added test cases to: LayoutTests/inspector/css/getComputedPrimaryFontForNode.html

* LayoutTests/inspector/css/getComputedPrimaryFontForNode.html:
* LayoutTests/inspector/css/getComputedPrimaryFontForNode-expected.txt:

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::buildObjectForFont):
- Send the already existing information about font synthesis to the frontend.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Font.js:
(WI.Font.fromPayload):
(WI.Font.prototype.get synthesizedBold):
(WI.Font.prototype.get synthesizedOblique):

* Source/WebInspectorUI/UserInterface/Views/DetailsSection.css:
(.details-section > .content > .group > .row.simple:has(.warning)):
(.details-section > .content > .group > .row.simple > .warning):
* Source/WebInspectorUI/UserInterface/Views/DetailsSectionSimpleRow.js:
(WI.DetailsSectionSimpleRow):
(WI.DetailsSectionSimpleRow.prototype.get warningMessage):
(WI.DetailsSectionSimpleRow.prototype.set warningMessage):
- Add support for putting the row in a warning state.

* Source/WebInspectorUI/UserInterface/Views/FontDetailsPanel.js:
(WI.FontDetailsPanel.prototype.refresh):
(WI.FontDetailsPanel.prototype.initialLayout):
- Adopt DetailsSectionWarnableSimpleRow and show warnings when the font is synthesized in some way.

Canonical link: https://commits.webkit.org/254188@main
